ddtrace/contrib/gevent/__init__.py

```diff
@@ -5,11 +5,8 @@
 The integration patches the gevent internals to add context management logic.
 
 .. note::
-    If :ref:`ddtrace-run<ddtracerun>` is being used set ``DD_GEVENT_PATCH_ALL=true`` and
-    ``gevent.monkey.patch_all()`` will be called as early as possible in the application
-    to avoid patching conflicts.
-    If ``ddtrace-run`` is not being used then be sure to call ``gevent.monkey.patch_all``
-    before importing ``ddtrace`` and calling ``ddtrace.patch`` or ``ddtrace.patch_all``.
+    If ``ddtrace-run`` is not being used then be sure to ``import ddtrace.bootstrap.sitecustomize``
```

Comment:
this is interesting, since the normal configuration we try to push people towards is `from ddtrace import patch_all; patch_all()`.

do these changes change that requirement and instead require that people import the sitecustomize module? (if yes, then I don't think I love the name very well, maybe we can create `ddtrace/auto.py` which is just `import ddtrace.bootstrap.sitecustomize`, this would match `import ddtrace.profiling.auto`)

Reply:
should we add a check to `ddtrace.bootstrap.sitecustomize` or `ddtrace.__init__` to check if `gevent` has been loaded before `ddtrace` and log a warning that you want to `import ddtrace` first? (or, should we have a check that the module cleanup has happened / we properly have the cloned modules we need?)

Reply:
My proposal would be to take the ddtrace-specific logic into `ddtrace.bootstrap.preload` and leave the boilerplate code in `sitecustomize.py`. Then users can just do
~~~
import ddtrace.bootstrap.preload  # noqa
~~~
Or we could have `ddtrace.bootstrap._preload` define the `preload` function that can be called by the boilerplate code in `sitecustomize.py`, and have users do
~~~
from ddtrace.bootstrap import preload

preload()
~~~

Reply:
Example here: https://github.com/DataDog/wilma/tree/main/wilma/_bootstrap

Reply:
but why should users have to import/call preload vs just importing the library?

is there not behavior there we want to occur for all usages of ddtrace? (specifically module cloning, there are probably some ddtrace-run specifics we don't want to happen on import of `ddtrace`)

Reply:
I suppose we _could_ ensure we perform the necessary steps on import, but then we would be relying on import side-effects. My feeling is that in general we tend to avoid those as much as possible.

Reply:
This PR should probably include new documentation about exactly how to use `ddtrace` with `gevent`. Including every step needed to run with `ddtrace-run` or manual patching with any requirements.

I feel like the one thing in this PR is "when using ddtrace-run it just works™️", but I want to make sure we have proper documentation/required steps for other cases as well since not everyone can/will use `ddtrace-run`.

Reply:
> My proposal would be to take the ddtrace-specific logic into `ddtrace.bootstrap.preload` and leave the boilerplate code in `sitecustomize.py`. 
@P403n1x87 @brettlangdon can we just do this but import this in `ddtrace.__init__.py` so that this will automatically run when users call `from ddtrace import patch_all`?

Reply:
@Yun-Kim I think that approach risks circular imports, but I could be wrong.

To @brettlangdon's question

> do these changes change that requirement and instead require that people import the sitecustomize module?

Yes, they do. I've committed a suggestion of how to handle this based on your first comment in this thread. https://github.com/DataDog/dd-trace-py/pull/5105/commits/81569f1f94b08a245c6fc9f84bdbc70b360ef91fEveryone paying attention, please let me know what you think.

Reply:
lgtm


I had a concern about `ddtrace.auto` (yes, I now I suggested the name) thinking that `ddtrace.tracing.auto` might be better, but our sitecustomize really is a central "any product enablement" activation, so I think it works!

a potential future improvement idea could be to try and isolate all the automated setup for each product in `ddtrace.{profiling,tracing,appsec,di}.auto` modules which get used in `sitecustomize.py`. could be nice way to segment them (neat idea, not a priority)